wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt

```diff
@@ -0,0 +1,210 @@
+
+    if (!extension.protoPaths.isEmpty() || !extension.protoTrees.isEmpty()) {
+      val allPaths = mergeDependencyPaths(project, extension.protoPaths, extension.protoTrees)
+      allPaths.forEach { path ->
+        protoConfiguration.dependencies.add(project.dependencies.create(path))
+      }
+    } else {
+      protoConfiguration.dependencies.addAll(sourceConfiguration.dependencies)
+    }
+
+    // at this point, all source and proto file references should be set up for Gradle to resolve.
+
+    val targets = mutableListOf<Target>()
+    val defaultBuildDirectory = "${project.buildDir}/generated/src/main/java"
+    val javaOutDirs = mutableListOf<String>()
+    val kotlinOutDirs = mutableListOf<String>()
+
+    val kotlinTarget = extension.kotlinTarget
+    val javaTarget = extension.javaTarget ?: if (kotlinTarget != null) null else JavaTarget()
+
+    javaTarget?.let { target ->
+      val javaOut = target.out ?: defaultBuildDirectory
+      javaOutDirs += javaOut
+      targets += Target.JavaTarget(
+          elements = target.elements ?: listOf("*"),
+          outDirectory = javaOut,
+          android = target.android,
+          androidAnnotations = target.androidAnnotations,
+          compact = target.compact
+      )
+    }
+    kotlinTarget?.let { target ->
+      val kotlinOut = target.out ?: defaultBuildDirectory
+      kotlinOutDirs += kotlinOut
+      targets += Target.KotlinTarget(
+          elements = target.elements ?: listOf("*"),
+          outDirectory = kotlinOut,
+          android = target.android,
+          javaInterop = target.javaInterop
+      )
+    }
+
+    val wireTask = project.tasks.register("generateProtos", WireTask::class.java) { task ->
+      task.source(sourceConfiguration)
+      task.sourceConfiguration = sourceConfiguration
+      task.protoConfiguration = protoConfiguration
+      task.roots = extension.roots?.asList() ?: emptyList()
+      task.prunes = extension.prunes?.asList() ?: emptyList()
+      task.rules = extension.rules
+      task.targets = targets
+      task.group = "wire"
+      task.description = "Generate Wire protocol buffer implementation for .proto files"
+    }
+
+    javaTarget?.let {
+      val compileTask = project.tasks.named("compileJava") as TaskProvider<JavaCompile>
+      compileTask.configure {
+        it.source(javaOutDirs)
+        it.dependsOn(wireTask)
+      }
+    }
+
+    kotlinTarget?.let {
+      val compileTask = project.tasks.named("compileKotlin") as TaskProvider<KotlinCompile>
+      compileTask.configure {
+        it.source(kotlinOutDirs)
+        it.dependsOn(wireTask)
+      }
+    }
+  }
+
+  private fun mergeDependencyPaths(
+    project: Project,
+    dependencyPaths: Set<String>,
+    dependencyTrees: Set<SourceDirectorySet> = emptySet()
+  ): List<Any> {
+    val allPaths = dependencyTrees.toMutableList<Any>()
+
+    dependencyPaths.forEach { path ->
+      val parser = FileOrUriNotationConverter.parser()
```

Comment:
nit: Is it possible to allocate the parser once vs on each iteration?

Reply:
oops, good catch.